### PR TITLE
perf: Use LinkedHashSet to deduplicate search domains

### PR DESCRIPTION
**Pull request by JunggiKim**
Motivation:

`DnsNameResolverBuilder.searchDomains` keeps unique domains in an `ArrayList`.
For every new domain, `List.contains` scans the values already collected. A large
list of distinct domains therefore performs a growing number of comparisons.

Modification:

Use one `LinkedHashSet` while reading the supplied `Iterable`. It removes
duplicates in expected constant time and preserves the first-seen order. Convert
the set to the existing `String[]` field after the loop, so resolver lookup order
and the stored representation stay unchanged.

Add a regression test for order, duplicate removal, case sensitivity, `Aa` / `BB`
hash collisions, and stopping at the first `null` value.

Result:

The duplicate-checking work changes from O(N^2) for N distinct domains to expected
O(N). A local JMH collection-loop benchmark on JDK 21, with 2 forks, measured:

| Distinct domains | ArrayList | LinkedHashSet |
| ---: | ---: | ---: |
| 64 | 3.196 us/op | 1.065 us/op |
| 256 | 46.678 us/op | 4.584 us/op |
| 1024 | 780.932 us/op | 26.025 us/op |
| 4096 | 12196.313 us/op | 147.224 us/op |

This is builder configuration work, not DNS request latency. Inputs containing
only repeated values may be slower because the old list stays at one element.

Verification:

```text
./mvnw -pl resolver-dns -am -Dtest=DnsNameResolverBuilderTest \
  -Dsurefire.failIfNoSpecifiedTests=false test
```

The reactor build succeeded. `DnsNameResolverBuilderTest` ran 10 tests with zero
failures, errors, or skipped tests.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@JunggiKim did you sign our icla yet ? https://netty.io/s/icla 

This change looks good... That said the impact is more or less meaningless in real-world as it is just used in the builder.

**Comment by JunggiKim** [comment]
@normanmaurer 
 Ah, I completely missed that! Thank you for kindly pointing it out—I just went ahead and signed it! 😊

Regarding the PR, while the performance impact in a real-world environment might be minimal, the change itself carries very low risk. Therefore, I believe it would still be a nice improvement to merge.

However, I completely respect your decision. If you feel this change isn't necessary, please feel free to close this PR (or let me know, and I can close it myself). Thanks again for your time and review!

**Comment by JunggiKim** [comment]
Hi @normanmaurer,

  I investigated the failed CI job.

  The only failed check is `linux-x86_64-java21-graal build`. It failed in
  `SingleThreadEventExecutorTest.testNotSuspendedUntilScheduledTaskIsCancelled()`
  from the `common` module.

  This PR changes only `resolver-dns/src/main/java/io/netty/resolver/dns/
  DnsNameResolverBuilder.java` and its corresponding DNS builder test. The failing
  test is unrelated to `DnsNameResolverBuilder`, `searchDomains()`, or this PR's
  `LinkedHashSet` change.

  TestLens marks this exact test as flaky: 4 flaky outcomes in 1,418 recent runs
  (about 1%). The failure is caused by a timing race while a cancelled scheduled task
  is being removed asynchronously from the event-loop queue. Netty previously
  addressed the same flaky test in #15256.

  Could a maintainer please rerun only the failed `linux-x86_64-java21-graal build`
  job?

  Thank you.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverBuilder.java [hunk 1]
```diff
@@ -557,22 +556,18 @@ public DnsNameResolverBuilder queryServerAddressStream(DnsServerAddressStream qu
     public DnsNameResolverBuilder searchDomains(Iterable<String> searchDomains) {
         checkNotNull(searchDomains, "searchDomains");
 
-        final List<String> list = new ArrayList<String>(4);
+        final LinkedHashSet<String> domains = new LinkedHashSet<String>(4);
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverBuilder.java]:
```suggestion
        final Set<String> domains = new LinkedHashSet<String>(4);
```

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverBuilder.java [hunk 2]
```diff
@@ -32,9 +32,8 @@
 
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.List;
+import java.util.LinkedHashSet;
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverBuilder.java:36]:
```suggestion
import java.util.LinkedHashSet;
import java.util.Set;
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by JunggiKim** [comment]
Thanks for applying the suggestions directly, @normanmaurer — using Set as the declared type is cleaner. Nothing further from my side the branch is up to date
with your commit. Let me know if you'd like anything else here.
